### PR TITLE
Replaced urllib2 with requests + added requests_cache which replaced Memoize and Lazy

### PR DESCRIPTION
**Pull request by norus**
Initial runtime in our case was about 10-15 minutes.

After replacing urllib2 with requests (and adding requests_cache on top), runtime substantially decreased.

### CODE REVIEW

File: puppet/puppet.py [hunk 1]
```diff
@@ -23,6 +29,9 @@ def certificate_requests(self):
             certnames.append(cert['name'])
         return certnames
 
+    def test(self):
```
Comment (daradib) [inline puppet/puppet.py]: Accidental commit?

File: puppet/puppet.py [hunk 2]
```diff
@@ -4,10 +4,16 @@
 class Puppet(object):
     """Python wrapper for Puppet REST API"""
 
-    def __init__(self, host='localhost', port=8140, key_file=None,
-        cert_file=None):
+    def __init__(self, host='localhost', port=8140, parser='yaml',
```
Comment (daradib) [inline puppet/puppet.py]: Move parser argument after cert_file to retain backwards-compatibility for positional arguments? Also update order in `README.md`.

**Comment by daradib** [comment]
Great work, thank you! Can you try running `test.py` if you haven't already (change hostnames of puppet master and node as appropriate)?

I think it might be better to default to in-memory caching, and require specifying `/tmp/pypuppet_cache.sqlite` explicitly. Thoughts?

**Comment by daradib** [comment]
Also, feel free to add yourself to the README. For example, you could change Author to a list of Contributors.

**Comment by norus** [comment]
Thanks for the detailed feedback!

I improved the code and here's the test result:

```
$ python -m doctest -v test.py
Trying:
    import puppet
Expecting nothing
ok
Trying:
    p = puppet.Puppet()
Expecting nothing
ok
Trying:
    p = puppet.Puppet(host='puppetmaster.foo.com',
    port=8140,
    key_file='/tmp/api-key.pem',
    cert_file='/tmp/api-cert.pem',
    ssl_verify=True,
    cache_enabled=True,
    cache_file='/tmp/pypuppet_cache',
    cache_backend='sqlite',
    cache_expire_after=3600)
Expecting nothing
ok
Trying:
    n = p.node('puppetnode.foo.com')
Expecting nothing
ok
Trying:
    str(n)
Expecting:
    'puppetnode.foo.com'
ok
Trying:
    dir(n)
Expecting:
    ['__doc__', '__init__', '__module__', '__str__', 'catalog', 'certificate', 'certificate_status', 'certname', 'classes', 'environment', 'facts', 'node', 'parameters', 'requestor']
ok
Trying:
    n_dev = p.node('puppetnode.foo.com', environment='production')
Expecting nothing
ok
Trying:
    n.certname
Expecting:
    'puppetnode.foo.com'
ok
Trying:
    type(n.classes)
Expecting:
    <type 'dict'>
ok
Trying:
    n.environment
Expecting:
    'production'
ok
Trying:
    type(n.facts)
Expecting:
    <type 'dict'>
ok
Trying:
    n.facts['osfamily'] + "-" + n.facts['architecture']
Expecting:
    'RedHat-x86_64'
ok
Trying:
    type(n.node)
Expecting:
    <type 'dict'>
ok
Trying:
    sorted(n.node.keys())
Expecting:
    ['classes', 'environment', 'expiration', 'facts', 'name', 'parameters', 'time']
ok
Trying:
    type(n.parameters)
Expecting:
    <type 'dict'>
ok
Trying:
    type(n.certificate())
Expecting:
    <type 'str'>
ok
Trying:
    n.certificate().startswith('-----BEGIN CERTIFICATE')
Expecting:
    True
ok
Trying:
    n.certificate_status()
Expecting:
    'signed'
ok
Trying:
    catalog = n.catalog()
Expecting nothing
ok
Trying:
    type(catalog)
Expecting:
    <type 'dict'>
ok
Trying:
    sorted(catalog.keys())
Expecting:
    [u'classes', u'edges', u'environment', u'name', u'resources', u'tags', u'version']
ok
Trying:
    [type(certnames[0]) for certnames in (p.certificates(),
    p.facts_search(('processorcount', 'ge', 1)))]
Expecting:
    [<type 'str'>, <type 'str'>]
ok
Trying:
    for certname in p.certificates():
        try:
            n_ = p.node(certname)
            break
        except puppet.APIError:
            # node probably does not exist
            continue
Expecting nothing
ok
Trying:
    dir(n_) == dir(n)
Expecting:
    True
ok
Trying:
    if p.facts_search(('architecture', 'amd64'),('osfamily', 'Debian')):
        print True
Expecting:
    True
ok
Trying:
    long_running_servers = p.facts_search(('uptime_days', 'gt', 100))
Expecting nothing
ok
Trying:
    crl = p.requestor.get('certificate_revocation_list', 'ca', parser='s')
Expecting nothing
ok
Trying:
    type(crl)
Expecting:
    <type 'str'>
ok
Trying:
    crl.startswith('-----BEGIN X509 CRL')
Expecting:
    True
ok
1 items passed all tests:
  29 tests in README.md
29 tests in 1 items.
29 passed and 0 failed.
Test passed.
1 items had no tests:
    test
0 tests in 1 items.
0 passed and 0 failed.
Test passed.
```

**Comment by norus** [comment]
Hi Dara,

Any issues?

**Comment by daradib** [comment]
My apologies for overlooking this for 4+ years. I've merged your changes, thanks! Hopefully this still works on newer versions of Puppet.

**Pull request merged**